garage/tf/models/base.py

```diff
@@ -117,32 +104,75 @@ def outputs(self):
 
     def __getstate__(self):
         args, kwargs = self.init_spec
-        return {
-            'args': args,
-            'kwargs': kwargs,
-            'weights': self.weights
-        }
+        return {'args': args, 'kwargs': kwargs, 'parameters': self.parameters}
 
     def __setstate__(self, data):
         # TODO(krzentner): There has got to be a better way.
         out = type(self)(*data['args'], **data['kwargs'])
-        out._default_weights = data['weights']
+        out._default_parameters = data['parameters']
         self.__dict__.update(out.__dict__)
 
 
 class TfModel(Model):
-
     def __init__(self, name):
         super().__init__()
         # All TF models need a valid unique name
         # TODO(ahtsan): produce a message on collision?
         self._name = name or self._name
+        self._is_parent = True
```

Comment:
Could you explain the design of parent, child, siblings, and twins to me, please? I can't seem to understand them after I reviewed all the code.

Reply:
### Design ###
It's a `Parent-Children` design, where parent model is the abstract model that holds the model/network structure, i.e. the variables, and the child model is spawned when given an input, to establish connectivity. This can also be interpreted as creating a symbolic model with the given input, using the same shared variables. It enforces 1-to-1 relationship in model and tf subgraph. Siblings and twins are not important.

### Usage ###
The suggested way to use the model is to first create a parent model (declare the structure of the model)
```python
model = SomeCustomModel()
```
and build the model with some given input.
The input and output of the first child model will be set as the default of the parent model.
```python
model.build(some_input_ph)
sess.run(model.outputs, feed_dict={model.inputs: some_data})
```

### Reusing the model for different input ###
This is the reason why we have child models. We can reuse the underlying variables by creating another child
```python
model.build(other_input_ph, name="another_child")
```
and access by
```python
model.another_child
```
### Scope ###
Unique name scope is required for each child model, and all of them share the same variable scope. So we cannot build a child model with the same name scope twice.
```python
model = SomeCustomModel()
model.build(some_input_ph) # it uses the default name scope
model.build(other_input_ph, name="another_child") # name scope here is "another_child"
model.build(other_input_ph) # error
```

### Pickling ###
Since the parent model contains the model/network structure, a.k.a the configuration, we only have to pickle the parent model. Therefore, child model are not supposed to be pickled.
Example:
```python
model = SomeCustomModel()
model_config = pickle.dumps(model)
...
model_restored = pickle.loads(model_config)
#establish connectivity in order to use it
model_restored.build(some_input_ph)
```

Reply:
Sorry I should have put some docstring.

Reply:
Wait, so passing in a name to the build method basically causes a copy of the graph to be made, but calling it normally builds it in place? Or do we end up with a mis-match between the model objects and the graph?

Either way, this seems quite surprising.

Reply:
see my comments on #512 on creating separate Graph an Model classes

In short, yes, but I think we should implement it to represent the individual Graphs and the Model (family of all graphs sharing variables) with separate classes.

I'm not sure it's really possible to improve on this, since this is exactly what TensorFlow does, and it's basically the only way to share variables.

I think the improvement here is to represent the process explicitly (TF makes does this silently and it's _very_ confusing as a newbie)

Anson has revised it so that calling `build` with no arguments creates a graph called "default".

Reply:
One alternative (which KR might prefer) is to treat `build` like a factory which returns a `Graph`. Then people can only interconnect `Graphs` in their code, not models (i.e. models would not have inputs and outputs).

Reply:
So as I understand it, the current design lets us do this:
```
model = SomeModel()
model.build(some_input)
sess.run(model.outputs, feed_dict={model.input: some_data})

other_output = model.build(some_other_input, name='copy1')
sess.run(other_output, feed_dict={some_other_input: some_data})
```

This allows having multiple copies of the subgraph that share variables, which is important only if you're going to assemble those copies and run them together. However, as I understand it, it's also possible to achieve this by just copying the variable around? I'm afraid I may not understand Tensorflow well enough to comprehend why we're doing this.

It's also still not clear to me what happens if I deserialize multiple instances of the `Model` object, and have them all build a default subgraph. Is that just broken? If so, can we detect this dynamically and throw an error?

But generally, Ryan's most recent comment seems most appealing to me, since it makes all of the subgraph instances "the same" instead of having one "default" subgraph instance, plus any number of "siblings."